File: src/main/java/main/commands/prefixcommands/audio/VolumeCommand.java
package main.commands.prefixcommands.audio;

import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.requestchannel.RequestChannelConfig;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;
import java.util.List;

@Deprecated
public class VolumeCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();
        final var guild = ctx.getGuild();

        if (ctx.getArgs().isEmpty()) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide what volume you'd like to set the bot to").build())
                    .queue();
            return;
        }

        if (!GeneralUtils.stringIsInt(ctx.getArgs().get(0))) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must provide an integer as the volume").build())
                    .queue();
            return;
        }

        final int volume = Integer.parseInt(ctx.getArgs().get(0));
        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();
        final GuildVoiceState selfVoiceState = ctx.getSelfMember().getVoiceState();

        msg.replyEmbeds(handleVolumeChange(selfVoiceState, memberVoiceState, volume).build())
                .queue();
    }

    public EmbedBuilder handleVolumeChange(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState, int volume) {
        final var guild = selfVoiceState.getGuild();

        if (!selfVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.VOICE_CHANNEL_NEEDED);

        if (memberVoiceState.inAudioChannel() && !memberVoiceState.getChannel().equals(selfVoiceState.getChannel()))
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));

        if (volume < 0 || volume > 100)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.VolumeMessages.INVALID_VOLUME);

        var musicManager = RobertifyAudioManager.getInstance().getMusicManager(memberVoiceState.getGuild());
        var audioPlayer = musicManager.getPlayer();

        audioPlayer.getFilters().setVolume((float)volume/100).commit();

        if (new RequestChannelConfig(guild).isChannelSet())
            new RequestChannelConfig(guild).updateMessage();

        new LogUtils(guild).sendLog(LogType.VOLUME_CHANGE,
                RobertifyLocaleMessage.VolumeMessages.VOLUME_CHANGED_LOG,
                Pair.of("{user}", memberVoiceState.getMember().getAsMention()),
                Pair.of("{volume}", String.valueOf(volume))
        );
        return RobertifyEmbedUtils.embedMessage(guild,
                RobertifyLocaleMessage.VolumeMessages.VOLUME_CHANGED,
                Pair.of("{volume}", String.valueOf(volume))
        );
    }

    @Override
    public String getName() {
        return "volume";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+GeneralUtils.listToString(getAliases())+"`\n\n" +
                "Control the volume of the bot\n\n" +
                "**__Usages__**\n`" +
                prefix + "volume <0-100>`";
    }

    @Override
    public List<String> getAliases() {
        return List.of("v", "vol");
    }

    @Override
    public boolean isPremiumCommand() {
        return true;
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/VolumeCommand.kt
package main.commands.slashcommands.audio

import dev.schlaubi.lavakord.audio.player.applyFilters
import main.audiohandlers.RobertifyAudioManager
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.utils.GeneralUtils.isNotNull
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.json.requestchannel.RequestChannelConfig
import main.utils.locale.messages.VolumeMessages
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType

class VolumeCommand : AbstractSlashCommand(
    SlashCommand(
        name = "volume",
        description = "Adjust the global volume of the bot",
        options = listOf(
            CommandOption(
                type = OptionType.INTEGER,
                name = "volume",
                description = "The volume to set the player to.",
                range = Pair(0, 200)
            )
        ),
        isPremium = true
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        event.replyEmbed {
            handleVolumeChange(
                memberVoiceState = event.member!!.voiceState!!,
                selfVoiceState = event.guild!!.selfMember.voiceState!!,
                volume = event.getRequiredOption("volume").asInt
            )
        }.queue()
    }

    private suspend fun handleVolumeChange(
        selfVoiceState: GuildVoiceState,
        memberVoiceState: GuildVoiceState,
        volume: Int
    ): MessageEmbed {
        val guild = selfVoiceState.guild
        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState, songMustBePlaying = true)
        if (acChecks.isNotNull()) return acChecks!!

        if (volume < 0 || volume > 200)
            return RobertifyEmbedUtils.embedMessage(guild, VolumeMessages.INVALID_VOLUME)
                .build()

        val player = RobertifyAudioManager[guild].player
        player.applyFilters {
            this.volume = volume / 100F
        }

        RequestChannelConfig(guild).updateMessage()

        LogUtilsKt(guild).sendLog(
            LogType.VOLUME_CHANGE,
            VolumeMessages.VOLUME_CHANGED_LOG,
            Pair("{user}", memberVoiceState.member.asMention),
            Pair("{volume}", volume.toString())
        )

        return RobertifyEmbedUtils.embedMessage(
            guild,
            VolumeMessages.VOLUME_CHANGED,
            Pair("{volume}", volume.toString())
        ).build()
    }

    override val help: String
        get() = """
                Control the volume of the bot

                **__Usages__**
                `/volume <0-100>`"""
}